core/parser.cpp

```diff
@@ -151,12 +157,13 @@ class Parser {
     /** Parse a comma-separated list of expressions.
      *
      * Allows an optional ending comma.
-     * \param exprs Expressions added here.
-     * \param end The token that ends the list (e.g. ] or )).
+     * \param args Expressions added here.
      * \param element_kind Used in error messages when a comma was not found.
+     * \param got_comma Whether a trailing comma was found.
+     * \param current_depth Current recursion depth to prevent stack overflow.
      * \returns The last token (the one that matched parameter end).
      */
-    Token parseArgs(ArgParams &args, const std::string &element_kind, bool &got_comma)
+    Token parseArgs(ArgParams &args, const std::string &element_kind, bool &got_comma, unsigned current_depth = 0)
```

Comment:
IMHO it's better not to provide a default value here. As far as I can tell there are only two call sites and both are changed in this PR to pass a current_depth value. And with a default of 0, forgetting to pass in the depth would be an easy way to introduce a code-path that can cause stack overflow.

The same comment applies for the other functions where you're adding a `current_depth` parameter; best not to give a default value for it.